### PR TITLE
fix(app): fix 404 page yields another 404

### LINKED ISSUE #4615: Clicking search suggestion on 404 page yields another 404
STR:

1. Go to a page that 404s (e.g. https://developer.mozilla.org/en-US/Using_native_JSON).
2. Type in a search query in the top right.
3. Click on one of the suggested results that pops up.

Actual:
You get another 404 page for the article linked from the suggested result (but it works if you refresh).

Expected:
You see the article that was suggested.

### PR DESCRIPTION
**Pull request by mustaphaturhan**
## Summary

Fixes #4615

### Problem

Clicking search suggestion on 404 page yields another 404 because of `useEffect` condition. Currently, `PageOrPageNotFound` component looking for `initialPathname.current && initialPathname.current !== pathname` to be true before `setNotFound(false)` but `initialPathName` and `pathname` always the same. It isn't triggering because of that.

### Solution

- Trigger setNotFound() after client renders the page.

## How did you test this change?

On my local environment. Steps:
1. Go somewhere which throws 404 like: `http://localhost:5042/en-US/Games/Tutorials/2D_Breakout_game_pure_JavaScript`
2. Search something from search bar and click one of the results.
2.1. Old behaviour was yields you to another 404.
2.1. New behaviour is showing correct page.

### CODE REVIEW

File: client/src/app.tsx [hunk 1]
```diff
@@ -112,6 +112,8 @@ function LoadingFallback({ message }: { message?: string }) {
 
 export function App(appProps) {
   const localeMatch = useMatch("/:locale/*");
+  const { pathname } = useLocation();
+  const initialPathname = React.useRef(pathname);
```
Comment (caugner) [inline client/src/app.tsx:116]: Was it necessary to move this up to the `App`?

File: client/src/app.tsx [hunk 2]
```diff
@@ -130,7 +132,10 @@ export function App(appProps) {
         <WritersHomepage />
       </Layout>
     ) : (
-      <PageOrPageNotFound pageNotFound={appProps.pageNotFound}>
+      <PageOrPageNotFound
+        pageNotFound={appProps.pageNotFound}
+        initialPathname={initialPathname}
```
Comment (caugner) [inline client/src/app.tsx:137]: Could we make the decision here by introducing a `pageNotFound` state that we set to false/true in a `useEffect()` that compares initialPathname and pathname and uses `appProps.pageNotFound` only if they're equal? This should allow us to eliminate the logic inside `PageOrPageNotFound`.

**Review by caugner — CHANGES_REQUESTED** [review]

**Review by caugner — APPROVED** [review]

File: client/src/app.tsx (hunk 2 above)
Comment (caugner) [reply client/src/app.tsx]: Went ahead and did the refactoring in 44086ba9f430ffa250e1097a6e4f5a477169d4c7 to land this.

**Comment by mustaphaturhan** [comment]
@caugner the changes you made are looks good to me. i think it is ready to go 🎉🎉🎉

**Comment by caugner** [comment]
Verified fixed on stage, thanks again! 🎉

**Comment by danielhjacobs** [comment]
This seems to have mostly fixed #6161 as well, though number 5 still applies: 

> Clicking back again gives the first 404ing page, but with a different error: "Loading Error

> Error: 404 on /en-US/docs/404/index.json: Page not found"

File: client/src/app.tsx [hunk 3]
```diff
@@ -119,6 +101,18 @@ export function App(appProps) {
...
+  const [pageNotFound, setPageNotFound] = React.useState<boolean>(
+    appProps.pageNotFound
+  );
+  const { pathname } = useLocation();
+  const initialPathname = React.useRef(pathname);
+
+  React.useEffect(() => {
+    if (initialPathname.current !== pathname) {
+      setPageNotFound(false);
+    }
```
Comment (caugner) [inline client/src/app.tsx:113]: @danielhjacobs This change might fix 5. of the referenced issue?

```suggestion
    setPageNotFound(initialPathname.current === pathname);
```

Comment (caugner) [reply client/src/app.tsx:113]: See: https://github.com/mdn/yari/pull/6459

**Pull request merged**